### ISSUE TITLE
Azure AD authentication error

### ISSUE THREAD

**Issue by eviweb** [maicol07/laravel-oidc-client#3]
Hi,
Authentication against Azure AD fails with this error [https://flareapp.io/share/o7AnDjJ5](https://flareapp.io/share/o7AnDjJ5)
The problem comes from that Azure needs the `grant_type` parameter to be set to `client_credentials` and passed in the body of the related request.
The authentication process is done here [https://github.com/maicol07/laravel-oidc-client/blob/ecb27a1a8214d6d8cde76fd41446b93b62aad7f7/src/Controllers/OIDCController.php#L37](https://github.com/maicol07/laravel-oidc-client/blob/ecb27a1a8214d6d8cde76fd41446b93b62aad7f7/src/Controllers/OIDCController.php#L37) and unfortunately I didn't find any obvious way to pass this setting.
Any idea ?

**Comment by eviweb** [comment]
Microsoft doc : [Get a token](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#get-a-token)

**Comment by maicol07** [comment]
Done! You can now use the new custom query parameters setting! Check it in the readme!

**Comment by eviweb** [comment]
Hi @maicol07,
unfortunately the `authorization_endpoint_query_params`, even if it works as expected, does not solve this issue as the `grant_type` parameter must be passed in the request **body** and not in the query string.
At this time, this problem appears a little bit more complex to me.
I think this issue should be re opened.
Best

**Comment by maicol07** [comment]
@eviweb oh wait, you're seeing OAuth2 docs. You must check the OIDC ones: [https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc.](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc.)

There are some differences between OAuth2 and OIDC (explained in that docs). Check if all the required parameters have been already fetched from the OIDC discovery JSON, otherwise you must override them manually via ENV.

Example for response\_type:  
Docs says the `response_type` must be set to `id_token`, so you might want to set `OIDC_RESPONSE_TYPES` to `id_token`.

I never used Azure AD, so I don't know much... It seems they're using a slightly different OIDC flow. However, you should be able to setup the Client with your configs. Let me know how it goes!

**Comment by eviweb** [comment]
@maicol07, I take just one second to tell you that you are right for the `response_type` parameter, many thanks.
And sorry for the confusion with the docs.
But the nightmare is not finished, I stay blocked on the Microsoft authentication screen even if my test user is logged.
There is a loop between the page and the callback URL.
I will continue to investigate and hope to find what's going wrong.
I will keep you informed.
If you'd like to check against Azure AD, Microsoft offers a complete environment for developers : https://developer.microsoft.com/en-us/microsoft-365/dev-program for free.

**Comment by eviweb** [comment]
Found, I need to add `'authorization_endpoint_query_params' => ['response_mode' => 'form_post'],` in the `oidc.php` config file.
Thanks to your last improvement.
One step further, we progress...

**Comment by eviweb** [comment]
Hi @maicol07,
the new problem I encounter is that the response sent by Azure AD uses the `POST` method and generates this error : [https://flareapp.io/share/17W6398P](https://flareapp.io/share/17W6398P).

I also have noticed that the `Client::getCurrentURL` method only relies on `$_SERVER['HTTPS']` and `$_SERVER['SERVER_PORT']` to detect the HTTP protocol [https://github.com/maicol07/oidc-client-php/blob/4d16b7be3dc01e4915f373979a1726435ca76be4/src/Client.php#L565](https://github.com/maicol07/oidc-client-php/blob/4d16b7be3dc01e4915f373979a1726435ca76be4/src/Client.php#L565), which should fail if the site is hosted by a web server listening on port 80 and set behind an SSL reverse proxy.
Maybe also checking `$_SERVER['HTTP_X_FORWARDED_PROTO']` and `$_SERVER['HTTP_X_FORWARDED_SSL']` variables would be more reliable.

**Comment by eviweb** [comment]
May I suggest to redefine the route `oidc.callback` as is:
```php
Route::match(['get', 'post'], config('oidc.callback_route_path'), [OIDCController::class, 'callback'])
        ->name('oidc.callback');
```

**Comment by maicol07** [comment]
@eviweb I agree adding the post method. Can you make a PR for that, referencing this issue?

When I have time I'll update the client to support all the known and additional flows of OIDC

**Comment by eviweb** [comment]
Hi @maicol07,
no problem for the PR.
I just tried the modification manually and now I'm stuck on the `oidc.callback` route with a page expired error.
I will continue to investigate but time is missing...
Any idea about this issue?
The exception seems to be thrown before the callback is called.

**Comment by maicol07** [comment]
@eviweb that's probably caused by Laravel CSRF verification. You can disable the middleware in that route:

```php
Route::match(['get', 'post'], config('oidc.callback_route_path'), [OIDCController::class, 'callback'])
		->withoutMiddleware('csrf')
        ->name('oidc.callback');
```

Or check if you can say to Azure to include the token in the callback

**Comment by eviweb** [comment]
Did not work using the `withoutMiddleware('crsf')` on the `oidc.callback` route, but adding an exception directly to the middleware class `VerifyCsrfToken` did the trick.
I'm now stuck in an infinite loop between the `authenticate` call from the `OIDCGuard::getUserInfo`.
_Edit: more precisely from the `Client::requestAuthorization` method._

**Comment by eviweb** [comment]
@maicol07 
I just create a draft PR for review.

**Comment by eviweb** [comment]
I succeed to go out from the loop to fall down in another issue, but I progress.

**Comment by maicol07** [comment]
@eviweb where are you stuck at the moment?

**Comment by eviweb** [comment]
@maicol07  I just find.
This is due to a mismatch between token signing algorithms, Azure AD provides `RS256` while the JWT package expects `HS256`.
I'm looking for a setting to specify the right algorithm.
May you help me please?

**Comment by eviweb** [comment]
@maicol07, 
how can I use `RSA` signing algorithm instead of `HMAC`?

**Comment by maicol07** [comment]
Oh yeah, sorry I've forgot about this. Currently, only Hmac works out-of-the-box in the Client: [https://github.com/maicol07/oidc-client-php/blob/11ee3e91cf26113c28c8b475737bcc007afa99e4/src/Client.php#L105](https://github.com/maicol07/oidc-client-php/blob/11ee3e91cf26113c28c8b475737bcc007afa99e4/src/Client.php#L105)

The JWT package supports RSA: [https://lcobucci-jwt.readthedocs.io/en/latest/configuration/#for-asymmetric-algorithms](https://lcobucci-jwt.readthedocs.io/en/latest/configuration/#for-asymmetric-algorithms)

I should update the Client to also support this. I have a few ideas, however I don't know when this will be implemented, since I'm a bit busy in this period. If you want to try to add that, this is my idea:

*   Add the types `HS256`, `HS512`, ... and RSA ones (`RS256`, `RS512`, ...). Since I want to be backward compatible, the old `sha256`, `sha512`, ... should default to `HMAC` signing.
*   Then detect if the chosen one uses HMAC or RSA
*   Finally, build the correct configuration object

Probably it will also require the private key.

Otherwise, I'll do it when I have a bit of time.

**Comment by maicol07** [comment]
@eviweb I've just pushed the update to support RSA:
- Client: https://github.com/maicol07/oidc-client-php/commit/0ee533edd6d5f5afbb17993cbcf03f860c6ce84d (v2.2)
- Laravel Client: https://github.com/maicol07/laravel-oidc-client/commit/a0b69965f5fa15770e46cbe99af14477402a9fca

**Comment by eviweb** [comment]
Hi @maicol07,
many thanks, I will give a try once the PR #5 will be merged.

**Comment by eviweb** [comment]
@maicol07,
I get the following error
```
Error
Typed property Maicol07\OpenIDConnect\Client::$jwt_signing_key must not be accessed before initialization
```
here: [https://github.com/maicol07/oidc-client-php/blob/577bcadd853bc33329b580d57c2e8e869b1a48ef/src/Traits/JWT.php#L75](https://github.com/maicol07/oidc-client-php/blob/577bcadd853bc33329b580d57c2e8e869b1a48ef/src/Traits/JWT.php#L75)
with _implicit flow_ on.
Sorry I can not share the stack trace on Flare (server issue).

The setting misses form the `$props` list in the constructor of the `Client` class.

**Comment by maicol07** [comment]
Fixed in https://github.com/maicol07/oidc-client-php/commit/ff2d24a636b6486b464a1757cdbcf14373b5f53d

Thank you for the report

**Comment by eviweb** [comment]
Too fast for me ;)

**Comment by eviweb** [comment]
So the PR I've done just add the setting in the related comment.

**Comment by maicol07** [comment]
What PR?

**Comment by eviweb** [comment]
I had prepare the same fix: [https://github.com/maicol07/oidc-client-php/pull/2](https://github.com/maicol07/oidc-client-php/pull/2).
But I had the settings in the comment block above your fix.

**Comment by eviweb** [comment]
@maicol07,
I encounter the following error:
```
Lcobucci\JWT\Signer\InvalidKeyProvided
It was not possible to parse your key, reason: error:0480006C:PEM routines::no start line
```
with the `SignedWith` constraint and find that `$signer->getPublicKey()` returns the private key.
I think this constraint should be initialized with the public key which should be used for validation and not the signing key.
IMHO, the `$config->signingKey()` should be replaced by `$config->verificationKey()` here,
[https://github.com/maicol07/oidc-client-php/blob/577bcadd853bc33329b580d57c2e8e869b1a48ef/src/Traits/JWT.php#L88](https://github.com/maicol07/oidc-client-php/blob/577bcadd853bc33329b580d57c2e8e869b1a48ef/src/Traits/JWT.php#L88)

**Comment by maicol07** [comment]
You're right! Changed in maicol07/oidc-client-php@fe0df96

**Comment by eviweb** [comment]
Hi @maicol07,
currently the `laravel-oidc-client` relies only on official releases of the `oidc-client-php` package, which prevents stable versions with bug fixes to be used before they are effectly released.
Maybe adding a branch alias would be a good idea, don't you think so?

Another question, is there a particular reason why the issue tracker is not opened for the `oidc-client-php` package?
I think that activating it would be more practical.
best

**Comment by maicol07** [comment]
> Currently the `laravel-oidc-client` relies only on official releases of the `oidc-client-php` package, which prevents stable versions with bug fixes to be used before they are effectly released.
> Maybe adding a branch alias would be a good idea, don't you think so?

You should be able to set dev-master as version constraint. However, I'll release the fix as a new version if there aren't issues

> Another question, is there a particular reason why the issue tracker is not opened for the `oidc-client-php` package?
> I think that activating it would be more practical.

Since it's a fork, issues are disabled by default. However, I'm thinking of moving to my own issue tracker at https://tracker.maicol07.it (has more options and I can organise things better than GitHub)

**Comment by eviweb** [comment]
> You should be able to set dev-master as version constraint. However, I'll release the fix as a new version if there aren't issues
This had to be done is the `composer.json` file of the `laravel-oidc-client` installed package, not in the Laravel site one, which is a little bit annoying.

**Comment by eviweb** [comment]
> You should be able to set dev-master as version constraint. However, I'll release the fix as a new version if there aren't issues

It generates a conflict of versions.
And finally find no way.

**Comment by maicol07** [comment]
@eviweb just released: https://github.com/maicol07/oidc-client-php/releases/tag/2.2.1

**Comment by eviweb** [comment]
Nice, thanks

**Comment by maicol07** [comment]
@eviweb I've setup the tracker. You can find this issue also on the tracker: [https://tracker.maicol07.it/issue/LARAOIDC-3](https://tracker.maicol07.it/issue/LARAOIDC-3)

Since the tracker automatically imported your user from Github, I don't know if you are able to login. If not, let me know and I'll find a way to get you in!

**Comment by eviweb** [comment]
Hi @maicol07,
thanks, unfortunately the Github authentication is not working.
Here is the URL I'm redirected to `https://bugs.maicol07.it/hub/api/rest/oauth2/interactive/login/f595197e-f6b9-4175-a3d5-d7978c09c0bb/land?error=redirect_uri_mismatch&error_description=The+redirect_uri+MUST+match+the+registered+callback+URL+for+this+application.&error_uri=https%3A%2F%2Fdocs.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-authorization-request-errors%2F%23redirect-uri-mismatch&state=bs86Cr3B`
Hope that helps.

**Comment by maicol07** [comment]
@eviweb now it should work. That was the old subdomain of the tracker

**Comment by eviweb** [comment]
It works like a charm ;)

**Comment by eviweb** [comment]
Hi @maicol07,
it seems your tracker license has expired.
I also want to discuss some improvements of the oidc client with you, what is the best channel for that?

**Comment by maicol07** [comment]
Hi,  
I'm renewing it, but unfortunately it's not instantly. You can chat with me on Telegram (you can check the link in my Github Profile readme)

**Comment by eviweb** [comment]
@maicol07 have you received my invitation on Telegram? It seems that I cannot contact you.

**Comment by maicol07** [comment]
Mmm... no. Try to search me (I'm `maicol07` even on Telegram)

**Comment by eviweb** [comment]
I found your account and wrote to you, but I got a message that indicates that due to your privacy policy I cannot send you any message...

**Comment by eviweb** [comment]
In case my Telegram username `eviweb`.

**Comment by maicol07** [comment]
Sent you a message
